api/callbacks.go

```diff
@@ -123,10 +123,10 @@ var db_vtable = C.DB_vtable{
 
 // contract: original pointer/struct referenced must live longer than C.DB struct
 // since this is only used internally, we can verify the code that this is the case
-func buildDB(kv KVStore, gm GasMeter) C.DB {
+func buildDB(kv *KVStore, gm *GasMeter) C.DB {
```

Comment:
KVStore is an interface and this is unneeded here.

Under the hood KVStore is `(type, pointer)` that points to the concrete implementation. The type info is used by reflect for dynamic dispatch.

So passing an interface "by value" is totally fine, it always uses a reference actually.

GasMeter is also an interface, I believe, so the above applies here as well. If GasMeter were a struct (not an interface pointing to a struct but a real struct) then you are totally write that would we be using a pointer to a function-local variable. In which case I would have assumed this to have blown up earlier.

Reply:
resolving following your new comment